### ISSUE TITLE
Re-do semantic information of atomic types in the database

### ISSUE THREAD

**Issue by thestr4ng3r** [radareorg/radare2#17649]
## Current situation

Currently when having atomic types in sdb, it looks like this:

```
float=type
type.float=f
type.float.size=32

double=type
type.double=F
type.double.size=64
```

the `type.float=f` part is the only thing that indicates how this type should be interpreted. However the `f` here is just the format for `pf` and it also embeds the size:

```
|  f       float value (4 bytes)
|  F       double value (8 bytes)
```

which is both redundant for the database and has the effect that you have to be aware of all the pf formats to detect that both `float` and `double` here are just floats.

## Solution

The database should just contain the basic info about how the type is supposed to be interpreted. The `type.%s` key can probably be used, but the values should be changed to something like this (just a proposal):

```
u=unsigned int
s=signed int
f=float
uc=unsigned char
sc=signed char
?=unknown
```

Note that `int`, `char` and `float` say nothing about their size, `wchar_t` could for example have `sc` + size=16.
These values should also be reflected in the `RAnalBaseType` struct in the form of an enum instead of passing the raw string.
Whenever the pf format is needed, it can be easily derived by this in combination with the size.

Imo pointer-types should not be saved in the database at all. They are merely pointers to some other base type (or void as a special case).

## Other implementations

For comparison, what it looks like in ghidra (also contains stuff like struct and array which makes no sense for us here):
https://github.com/thestr4ng3r/ghidra/blob/7bde3b54b43230601363f89b0214ab4bdba8bf6f/Ghidra/Features/Decompiler/src/decompile/cpp/type.hh#L33-L46
https://github.com/thestr4ng3r/ghidra/blob/7bde3b54b43230601363f89b0214ab4bdba8bf6f/Ghidra/Features/Decompiler/src/decompile/cpp/type.hh#L164-L178
Note that the base type here is really just a tuple `int4 s, type_metatype m, const string &n`

And the respective part of dwarf (see http://dwarfstd.org/doc/DWARF5.pdf):
![20200914_11h45m21s_grim](https://user-images.githubusercontent.com/1460997/93071009-db5ff580-f67f-11ea-9f30-fa865ae0716e.png)

**Comment by XVilka** [comment]
Regarding these:
```
u=unsigned int
s=signed int
f=float
uc=unsigned char
sc=signed char
?=unknown
```
I think they are not "atomic" types since they heavily dependent on the target architecture, operating system, bitness, phase of the moon, and a compiler authors' whim. Lets stick to the standard `uint8_t`, `uint16_t`, etc.

**Comment by thestr4ng3r** [comment]
I addressed this explicitly in the description but maybe it was not clear enough:
> Note that int, char and float say nothing about their size, wchar_t could for example have sc + size=16.

Here `u` in the sense of `u=unsigned int` means only "some integer value interpreted as unsigned" with no info about the size and not directly related to `unsigned int` in C. The size is purely defined by the extra `size` field, hence why there is no `unsigned short`, `unsigned long`, etc. in this list.

In the end, you will have a database (loaded dynamically depending on compiler/os/whatever) with info that would look like this:
```
unsigned int = u, size 4
unsigned long = u, size 4
unsigned long long = u, size 8
uint32_t = u, size 4
int = s, size 4
...
```

**Comment by HoundThe** [comment]
I am not completely sure I understand the issue. Instead of having a formatting character that represents an encoding *and* size, have just an encoding, as size is a separate attribute of the type?

**Comment by thestr4ng3r** [comment]
Yes exactly.
Before it was: "encoding+size" formatting char + size
After it will be: encoding enum + size

**Comment by HoundThe** [comment]
Yeah, I agree that sounds better and more flexible for the future, there are few encodings but a lot of different sizes.
